src/VCR/Cassette.php

```diff
@@ -56,26 +56,43 @@ public function __construct($name, Configuration $config, Storage $storage)
      */
     public function hasResponse(Request $request)
     {
-        return $this->playback($request) !== null;
+        try {
+            $response = $this->playback($request);
+            return $response !== null;
+        } catch (\OutOfBoundsException $e) {
+            return false;
+        }
     }
 
     /**
      * Returns a response for given request or null if not found.
      *
      * @param Request $request Request.
      *
-     * @return Response|null Response for specified request.
+     * @return Response Response for specified request.
+     *
+     * @throws \OutOfBoundsException
+     *   Thrown if the cassette does not contain the response.
      */
     public function playback(Request $request)
     {
+        $storedRequestData = array();
         foreach ($this->storage as $recording) {
             $storedRequest = Request::fromArray($recording['request']);
             if ($storedRequest->matches($request, $this->getRequestMatchers())) {
                 return Response::fromArray($recording['response']);
             }
+            $storedRequestData[] = $storedRequest->toArray();
```

Comment:
$storedRequestData could grow very large (in particular if we are dealing with many file uploads).

This could lead to memory issues, and the "Stored requests:" part of the exception will very likely be not that readable.

Reply:
In my experience having values from the stored requests and the used matchers is key to determining why none of the stored requests apply when you think they should. This includes the url, headers etc.

I agree about problems regarding readability and file uploads. Perhaps some fields could be excluded or truncated.